File: backend/app/api/v1/endpoints/api_keys.py
"""
API Key Management Endpoints
Allows users to generate and manage API keys with rotation policies
"""

from typing import Annotated, List, Optional
from datetime import datetime
from fastapi import APIRouter, Depends, HTTPException, Request, status
from sqlalchemy.ext.asyncio import AsyncSession
from pydantic import BaseModel, Field

from app.core.database import get_db
from app.core.rate_limit import rate_limit_decorator
from app.core.security_audit import SecurityAuditLogger, SecurityEventType
from app.models.user import User
from app.models.api_key import APIKey
from app.api.v1.endpoints.auth import get_current_user
from app.services.api_key_service import APIKeyService, APIKeyRotationPolicy
from app.dependencies import require_admin_or_superadmin
from sqlalchemy import select
from sqlalchemy.orm import selectinload

router = APIRouter()


class APIKeyCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    description: Optional[str] = Field(None, max_length=1000)
    rotation_policy: str = Field(
        default="manual",
        description="Rotation policy: manual, 30d, 60d, 90d, 180d, 365d"
    )
    expires_in_days: Optional[int] = Field(None, ge=1, le=3650, description="Expiration in days (optional)")


class APIKeyResponse(BaseModel):
    id: int
    name: str
    key: str  # Only shown once on creation
    key_prefix: str
    created_at: str
    expires_at: Optional[str] = None
    last_used_at: Optional[str] = None
    rotation_policy: str
    next_rotation_at: Optional[str] = None


class APIKeyListResponse(BaseModel):
    id: int
    name: str
    key_prefix: str
    created_at: str
    expires_at: Optional[str] = None
    last_used_at: Optional[str] = None
    rotation_policy: str
    next_rotation_at: Optional[str] = None
    rotation_count: int
    usage_count: int
    is_active: bool


class AdminAPIKeyListResponse(BaseModel):
    id: int
    name: str
    key_prefix: str
    created_at: str
    expires_at: Optional[str] = None
    last_used_at: Optional[str] = None
    rotation_policy: str
    next_rotation_at: Optional[str] = None
    rotation_count: int
    usage_count: int
    is_active: bool
    user_id: int
    user_email: str
    user_name: Optional[str] = None


class APIKeyRotateResponse(BaseModel):
    old_key_id: int
    new_key: APIKeyResponse
    message: str


@router.post("/generate", response_model=APIKeyResponse)
@rate_limit_decorator("5/minute")
async def generate_api_key_endpoint(
    request: Request,
    api_key_data: APIKeyCreate,
    current_user: Annotated[User, Depends(get_current_user)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    """Generate a new API key for the current user"""
    # Validate rotation policy
    if not APIKeyRotationPolicy.is_valid_policy(api_key_data.rotation_policy):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Invalid rotation policy. Valid options: {', '.join(APIKeyRotationPolicy.POLICIES.keys())}"
        )
    
    # Create API key
    api_key_model, plaintext_key = await APIKeyService.create_api_key(
        db=db,
        user=current_user,
        name=api_key_data.name,
        description=api_key_data.description,
        rotation_policy=api_key_data.rotation_policy,
        expires_in_days=api_key_data.expires_in_days,
    )
    
    # Log security event
    await SecurityAuditLogger.log_api_key_event(
        db=db,
        event_type=SecurityEventType.API_KEY_CREATED,
        api_key_id=api_key_model.id,
        description=f"API key '{api_key_data.name}' created",
        user_id=current_user.id,
        user_email=current_user.email,
        ip_address=request.client.host if request.client else None,
        metadata={
            "rotation_policy": api_key_data.rotation_policy,
            "expires_in_days": api_key_data.expires_in_days,
        }
    )
    
    return APIKeyResponse(
        id=api_key_model.id,
        name=api_key_model.name,
        key=plaintext_key,  # Only time this is shown
        key_prefix=api_key_model.key_prefix,
        created_at=api_key_model.created_at.isoformat(),
        expires_at=api_key_model.expires_at.isoformat() if api_key_model.expires_at else None,
        last_used_at=api_key_model.last_used_at.isoformat() if api_key_model.last_used_at else None,
        rotation_policy=api_key_model.rotation_policy,
        next_rotation_at=api_key_model.next_rotation_at.isoformat() if api_key_model.next_rotation_at else None,
    )


@router.get("/list", response_model=List[APIKeyListResponse])
@rate_limit_decorator("10/minute")
async def list_api_keys(
    request: Request,
    current_user: Annotated[User, Depends(get_current_user)],
    db: Annotated[AsyncSession, Depends(get_db)],
    include_inactive: bool = False,
):
    """List all API keys for the current user"""
    api_keys = await APIKeyService.get_user_api_keys(
        db=db,
        user=current_user,
        include_inactive=include_inactive,
    )
    
    return [
        APIKeyListResponse(
            id=key.id,
            name=key.name,
            key_prefix=key.key_prefix,
            created_at=key.created_at.isoformat(),
            expires_at=key.expires_at.isoformat() if key.expires_at else None,
            last_used_at=key.last_used_at.isoformat() if key.last_used_at else None,
            rotation_policy=key.rotation_policy,
            next_rotation_at=key.next_rotation_at.isoformat() if key.next_rotation_at else None,
            rotation_count=key.rotation_count,
            usage_count=key.usage_count,
            is_active=key.is_active,
        )
        for key in api_keys
    ]


@router.post("/{key_id}/rotate", response_model=APIKeyRotateResponse)
@rate_limit_decorator("5/minute")
async def rotate_api_key(
    request: Request,
    key_id: int,
    current_user: Annotated[User, Depends(get_current_user)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    """Rotate an API key (creates new key, deactivates old one)"""
    try:
        new_key_model, plaintext_key = await APIKeyService.rotate_api_key(
            db=db,
            api_key_id=key_id,
            user=current_user,
        )
        
        # Log security event
        await SecurityAuditLogger.log_api_key_event(
            db=db,
            event_type=SecurityEventType.API_KEY_ROTATED,
            api_key_id=new_key_model.id,
            description=f"API key rotated (old: {key_id}, new: {new_key_model.id})",
            user_id=current_user.id,
            user_email=current_user.email,
            ip_address=request.client.host if request.client else None,
            metadata={
                "old_key_id": key_id,
                "new_key_id": new_key_model.id,
            }
        )
        
        return APIKeyRotateResponse(
            old_key_id=key_id,
            new_key=APIKeyResponse(
                id=new_key_model.id,
                name=new_key_model.name,
                key=plaintext_key,  # Only time this is shown
                key_prefix=new_key_model.key_prefix,
                created_at=new_key_model.created_at.isoformat(),
                expires_at=new_key_model.expires_at.isoformat() if new_key_model.expires_at else None,
                last_used_at=new_key_model.last_used_at.isoformat() if new_key_model.last_used_at else None,
                rotation_policy=new_key_model.rotation_policy,
                next_rotation_at=new_key_model.next_rotation_at.isoformat() if new_key_model.next_rotation_at else None,
            ),
            message="API key rotated successfully. Old key has been deactivated.",
        )
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(e)
        )


@router.delete("/{key_id}")
@rate_limit_decorator("10/minute")
async def revoke_api_key(
    request: Request,
    key_id: int,
    current_user: Annotated[User, Depends(get_current_user)],
    db: Annotated[AsyncSession, Depends(get_db)],
    reason: Optional[str] = None,
):
    """Revoke an API key"""
    try:
        api_key = await APIKeyService.revoke_api_key(
            db=db,
            api_key_id=key_id,
            user=current_user,
            reason=reason,
        )
        
        # Log security event
        await SecurityAuditLogger.log_api_key_event(
            db=db,
            event_type=SecurityEventType.API_KEY_REVOKED,
            api_key_id=api_key.id,
            description=f"API key '{api_key.name}' revoked",
            user_id=current_user.id,
            user_email=current_user.email,
            ip_address=request.client.host if request.client else None,
            metadata={
                "reason": reason,
            }
        )
        
        return {"message": "API key revoked successfully"}
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(e)
        )


@router.get("/admin/list", response_model=List[AdminAPIKeyListResponse])
@rate_limit_decorator("20/minute")
async def admin_list_all_api_keys(
    request: Request,
    current_user: Annotated[User, Depends(get_current_user)],
    db: Annotated[AsyncSession, Depends(get_db)],
    _: None = Depends(require_admin_or_superadmin),
    include_inactive: bool = False,
    user_id: Optional[int] = None,
):
    """List all API keys (admin only). Can filter by user_id."""
    # Build query
    query = select(APIKey)
    
    if user_id:
        query = query.where(APIKey.user_id == user_id)
    
    if not include_inactive:
        query = query.where(APIKey.is_active == True)
    
    query = query.order_by(APIKey.created_at.desc())
    
    # Load user relationship
    query = query.options(selectinload(APIKey.user))
    
    result = await db.execute(query)
    api_keys = result.scalars().all()
    
    return [
        AdminAPIKeyListResponse(
            id=key.id,
            name=key.name,
            key_prefix=key.key_prefix,
            created_at=key.created_at.isoformat(),
            expires_at=key.expires_at.isoformat() if key.expires_at else None,
            last_used_at=key.last_used_at.isoformat() if key.last_used_at else None,
            rotation_policy=key.rotation_policy,
            next_rotation_at=key.next_rotation_at.isoformat() if key.next_rotation_at else None,
            rotation_count=key.rotation_count,
            usage_count=key.usage_count,
            is_active=key.is_active,
            user_id=key.user_id,
            user_email=key.user.email if key.user else "Unknown",
            user_name=f"{key.user.first_name or ''} {key.user.last_name or ''}".strip() if key.user else None,
        )
        for key in api_keys
    ]

File: apps/web/src/lib/api/apiKeys.ts
/**
 * API Keys API client
 * Manages API key creation, listing, rotation, and revocation
 */

import { apiClient } from './client';
import { extractApiData } from './utils';

export interface APIKeyCreate {
  name: string;
  description?: string;
  rotation_policy?: 'manual' | '30d' | '60d' | '90d' | '180d' | '365d';
  expires_in_days?: number;
}

export interface APIKeyResponse {
  id: number;
  name: string;
  key: string; // Only shown once on creation
  key_prefix: string;
  created_at: string;
  expires_at?: string | null;
  last_used_at?: string | null;
  rotation_policy: string;
  next_rotation_at?: string | null;
}

export interface APIKeyListResponse {
  id: number;
  name: string;
  key_prefix: string;
  created_at: string;
  expires_at?: string | null;
  last_used_at?: string | null;
  rotation_policy: string;
  next_rotation_at?: string | null;
  rotation_count: number;
  usage_count: number;
  is_active: boolean;
}

export interface APIKeyRotateResponse {
  old_key_id: number;
  new_key: APIKeyResponse;
  message: string;
}

export interface AdminAPIKeyListResponse extends APIKeyListResponse {
  user_id: number;
  user_email: string;
  user_name?: string | null;
}

/**
 * API Keys API client
 */
const apiKeysAPI = {
  /**
   * Generate a new API key
   */
  create: async (data: APIKeyCreate): Promise<APIKeyResponse> => {
    const response = await apiClient.post<APIKeyResponse>('/v1/api-keys/generate', data);
    const result = extractApiData<APIKeyResponse>(response);
    if (!result) {
      throw new Error('Failed to create API key: no data returned');
    }
    return result;
  },

  /**
   * List all API keys for the current user
   */
  list: async (includeInactive: boolean = false): Promise<APIKeyListResponse[]> => {
    const response = await apiClient.get<APIKeyListResponse[]>('/v1/api-keys/list', {
      params: { include_inactive: includeInactive },
    });
    const data = extractApiData<APIKeyListResponse[] | { items: APIKeyListResponse[] }>(response);
    if (Array.isArray(data)) {
      return data;
    }
    if (data && typeof data === 'object' && 'items' in data) {
      return (data as { items: APIKeyListResponse[] }).items;
    }
    return [];
  },

  /**
   * Rotate an API key (creates new key, deactivates old one)
   */
  rotate: async (keyId: number): Promise<APIKeyRotateResponse> => {
    const response = await apiClient.post<APIKeyRotateResponse>(`/v1/api-keys/${keyId}/rotate`);
    const result = extractApiData<APIKeyRotateResponse>(response);
    if (!result) {
      throw new Error('Failed to rotate API key: no data returned');
    }
    return result;
  },

  /**
   * Revoke an API key
   */
  revoke: async (keyId: number, reason?: string): Promise<void> => {
    await apiClient.delete(`/v1/api-keys/${keyId}`, {
      params: reason ? { reason } : undefined,
    });
  },

  /**
   * List all API keys (admin only)
   */
  adminList: async (includeInactive: boolean = false, userId?: number): Promise<AdminAPIKeyListResponse[]> => {
    const response = await apiClient.get<AdminAPIKeyListResponse[]>('/v1/api-keys/admin/list', {
      params: {
        include_inactive: includeInactive,
        ...(userId && { user_id: userId }),
      },
    });
    const data = extractApiData<AdminAPIKeyListResponse[] | { items: AdminAPIKeyListResponse[] }>(response);
    if (Array.isArray(data)) {
      return data;
    }
    if (data && typeof data === 'object' && 'items' in data) {
      return (data as { items: AdminAPIKeyListResponse[] }).items;
    }
    return [];
  },
};

export default apiKeysAPI;
export { apiKeysAPI };
